src/storage-client/src/controller/rehydration.rs

```diff
@@ -239,63 +245,48 @@ where
-                                Ok(client) => Ok(client),
-                                Err(e) => {
-                                    if state.i >= mz_service::retry::INFO_MIN_RETRIES {
-                                        tracing::info!(
-                                        "error connecting to {:?} for storage, retrying in {:?}: {e}",
-                                        location,
-                                        state.next_backoff.unwrap()
-                                    );
-                                    } else {
-                                        tracing::debug!(
-                                        "error connecting to {:?} for storage, retrying in {:?}: {e}",
-                                        location,
-                                        state.next_backoff.unwrap()
-                                    );
-                                    }
-                                    Err(e)
-                                }
-                            }?;
-
-                        client
-                            .send(StorageCommand::CreateTimely {
-                                config: timely_config,
-                                epoch: ClusterStartupEpoch::new(envd_epoch, 0),
-                            })
-                            .await?;
-
-                        Ok::<_, anyhow::Error>(client)
+            let dests = location
+                .ctl_addrs
+                .clone()
+                .into_iter()
+                .map(|addr| (addr, self.metrics.clone()))
+                .collect();
+            let version = self.build_info.semver_version();
+            let client = StorageGrpcClient::connect_partitioned(dests, version).await;
+
+            let client = match client {
+                Ok(client) => client,
+                Err(e) => {
+                    if state.i >= mz_service::retry::INFO_MIN_RETRIES {
+                        tracing::info!(
+                            "error connecting to {:?} for storage, retrying in {:?}: {e}",
+                            location,
+                            state.next_backoff.unwrap()
+                        );
+                    } else {
+                        tracing::debug!(
+                            "error connecting to {:?} for storage, retrying in {:?}: {e}",
+                            location,
+                            state.next_backoff.unwrap()
+                        );
                     }
-                })
-                .await
-                .expect("retry retries forever");
+                    continue;
+                }
+            };
+
+            let timely_command = StorageCommand::CreateTimely {
+                config: timely_config,
+                // The replica epoch is unused by storage, and won't
+                // be used till multi-replica storage is supported.
+                epoch: ClusterStartupEpoch::new(envd_epoch, 0),
```

Comment:
Don't we need to count up the replica epoch even with just a single replica? It is used to ensure that the connections between processes get initialized correctly and we don't run into dead-locks when a strict subset of the replica processes crashes during initialization. @umanwizard has the most context on what can go wrong.

Reply:
Yes, good catch. This number needs to increase every time we attempt to launch the same replica.

Reply:
This epoch is _never_ checked in the storage code, its only checked in the `mz_compute_client::controller::{instance, replica}` modules, but i can bump it in this pr for clarity sake

Reply:
The point is that the storage code *should* probably check the epoch if we want to support multi-process storage clusters, to avoid running into the same kind of split-brain issues that Brennan had to deal with before he added that epoch.

The comment makes it sound like the epoch is needed to support multi-replica clusters, but that's not the case. Multiple replicas can get by fine without that.

Reply:
> This epoch is _never_ checked in the storage code

Then what are you filling in the `CreateTimely` command that gets sent to clusterd (and ultimately ends up in `communication.rs` ?

Reply:
It doesn't need to be checked in any of the envd-side code; it needs to be properly maintained on that side, and then used in communication.rs

Reply:
As for what can go wrong otherwise, nobody really knows in detail, there are just a million difficult-to-reason-about edge cases. That’s why I ripped out the old logic and replaced it with this epoch checking.

But the main theme is that in a multi-process cluster, if some processes fail to connect and others succeed, the failed ones might be rebooted and respond to a _future_ CreateTimely command. Then you get into a situation where some of the processes have responded to one command and others to another. That’s why now we made them check that they all have the same epoch in CreateTimely and otherwise refuse to let them build a timely instance together. So, every CreateTimely that is instantiated for a given replica needs to have a strictly increasing epoch.